### PR TITLE
Subnautica: correctly handle unhatchable aggressive species

### PR DESCRIPTION
**Pull request by Berserker66**

### CODE REVIEW

File: worlds/subnautica/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import functools
```
Comment (black-sliver) [inline worlds/subnautica/__init__.py:1]: This is unused by the code below?

**Review by black-sliver — COMMENTED** [review]
Hm, the .Cache is kind of awkward. Maybe call it `Definitions` or something and move the unsorted list in as well? Or you could use module-level getattr to provide the cached ones on module level, i think.

File: worlds/subnautica/__init__.py (hunk 1 above)
Comment (Berserker66) [reply worlds/subnautica/__init__.py:1]: the decorator is @functools.cached_property

Comment (Berserker66) [reply worlds/subnautica/__init__.py:1]: oh its in init still.. yeah, that's where the Cache first existed before I moved it.

Comment (Berserker66) [reply worlds/subnautica/__init__.py:1]: > Hm, the .Cache is kind of awkward. Maybe call it `Definitions` or something and move the unsorted list in as well? Or you could use module-level getattr to provide the cached ones on module level, i think.

A module-level wrapper to cache on access could maybe be a generally available Utils class/type?

**Review by recklesscoder — COMMENTED** [review]
I think this code works as expected.

Not directly related to these changes, I noticed the following creatures are missing from the aggressive creature list:
* Stalkers
* Sand Sharks
* Biters
* Blighters
* Blood Crawlers

And finally, I'm not sure whether this is in scope for this changeset, but here goes:
As there is a huge difference in difficulty in scanning, say, a Bleeder and a Reaper Leviathan, which are currently classified exactly the same (as "aggressive" and "not hatchable"), maybe there should be options available that differentiate these two cases.
Further, most aggressive creatures are trivial to scan without Stasis Rifle or Alien Containment by anyone who has played through Subnautica once or twice (or perhaps a few times) by simply circle-strafing around the creature while maybe tanking a few hits. The only exceptions I see where an experienced player might desire a Stasis Rifle are the aggressive Leviathans, the Warper and the Crashfish.

So I propose splitting the `creature_scan_logic` option into these three more flexible options, disentangling the semantics a bit in the process:
```yaml
aggressive_creature_scan_size: # If creature_scans are enabled, what aggressive creatures to consider.
  #    None: No items are placed on aggressive creatures.
  #    Small: Include small creatures such as Biters.
  #    Medium: Include all creatures below leviathan class.
  #    All: Include all creatures.
  none: 0
  small: 0
  medium: 0
  all: 50
aggressive_creature_scan_stasisrifle: # If creature_scans are enabled, which aggressive
  #    creatures are considered unscannable until the Stasis Rifle is unlocked.
  #    All: Even the smallest aggressive creatures are deemed to require a Stasis Rifle.
  #    Medium: Anything larger than a Biter is deemed to require a Stasis Rifle.
  #    Difficult: Only leviathans plus Warpers and Crashfish are deemed to require a Stasis Rifle.
  #    None: Who needs a Stasis Rifle? Surely the player will come up with something.
  all: 0
  medium: 50
  difficult: 0
  none: 0
creature_scans_from_containment: # If creature_scans are enabled, whether using
  #    Alien Containment to obtain creature scans is considered.
  #    False: All required scans are possible without unlocking Alien Containment.
  #    True: Some aggressive creatures that would require a Stasis Rifle (according to
  #          the aggressive_creature_scan_stasisrifle option) may instead be expected
  #          to be hatched in an Alien Containment tank instead.
  #          This setting also unlocks the Cuddlefish scan as a possible item location.
  # 
  false: 50
  true: 0
```

Notes:
* This proposal can recreate the logic of all current `creature_scan_logic` values except for `containment`. I'm not sure that value would have found much use, as scanning directly (without a Stasis Rifle) is usually a lot faster and easier than finding and hatching an egg. Perhaps one could add hatching-based locations instead.
* Technically the first of these three options can already be emulated by listing undesired creatures under `exclude_locations` (with minor semantic differences), as can `creature_scan_logic: removed`. However, I believe many players would appreciate this shorthand to avoid signing up for Reaper scanning duty.
* The line between "small" and "medium" aggressive creatures is not arbitrary: it correlates with the danger posed as the small creatures all do 5–10 damage per hit, while the medium creatures all do 30–50 damage per hit.

If time is pressing and you want to go ahead ignoring this proposal, I wouldn't mind moving this to an issue for further discussion.

**Comment by Berserker66** [comment]
I'm curious how I missed the other aggressives, I must have gotten distracted with something and left the list incomplete.

I'm not all that happy with the big switch option with many exceptions, but I'm also not yet happy with any of the alternatives. Something I tried was splitting the Range in 3, having passive pool, stasis pool and containment pool; as well as a toggle for logic on stasis and a toggle for cuddlefish. I wasn't happy with that solution and I find yours less appealing.

Time is technically pressing as this PR already missed the RC1 window.

**Comment by black-sliver** [comment]
> So I propose splitting the creature_scan_logic option into these three more flexible options,

(In other games) we often collect items "in logic" and "out of logic". This can be applied to "soft progression" as well: you can chose to scan something you are not "supposed to" and maybe take a shortcut to the goal, or not. What you suggest is similar to having "glitches in logic" where the player can configure what is considered to be in logic (while not in logic from a vanilla standpoint). I think that is only appealing if it generates vastly different playthroughs, or when it really changes difficulty of a game (so you can have "easy" and "hard mode") - archipelago is not really equipped to force a hard seed (yet), so the result would still be random.

**Review by black-sliver — APPROVED** [review]
From the conversation, I think all actual bugs have been addressed.
Code lgtm and rolling works. In case you haven't, I would suggest loading/connecting a game or two and checking a spoiler log or two before hitting merge.

**Comment by recklesscoder** [comment]
I think my main gripe is that if I want a Stasis Rifle for the few "difficult" scans, I need to use settings that assume I need it for a whole host of trivial scans, leading to my games probably always giving me a Stasis Rifle early. Maybe that fear is unfounded; I'll continue keeping an eye on the Stasis Rifles in my seeds. @black-sliver also makes a good point in that if these settings don't cause vastly different playthroughs, they are really mostly cosmetic. Getting Stasis Rifle every game doesn't hurt if you barely use it.

As for the "vanilla standpoint", from my experience it is extremely common for first-time players to scan the small creatures as well as Stalkers and Sand Sharks, which you meet early in the game, without a Stasis Rifle. On the other hand, certain creatures are almost certainly intended to be scanned using a Stasis Rifle, as they are near impossible to get a hold of without. In that way, when compared to vanilla, our logic is either overbearingly protective or ridiculously hands-off.

Anyhow, I can live with this for now. Maybe it's just me and it won't bother anyone else.

**Pull request merged**